boards/support/tock_rubble/src/tock_rubble.rs

```diff
@@ -0,0 +1,449 @@
+            )),
+        ))
+    }
+
+    fn has_work(&mut self) -> bool {
+        self.0.has_work()
+    }
+    fn process_one(&mut self) -> Result<(), Self::Error> {
+        self.0.process_one()
+    }
+}
+
+pub struct LinkLayerWrapper<'a, A, R>(rubble::link::LinkLayer<RubbleConfig<'a, A, R>>)
+where
+    A: time::Alarm<'a>,
+    R: hil::rubble::radio::RubbleData<'a>
+        + ble_advertising::BleAdvertisementDriver<'a>
+        + ble_advertising::BleConfig
+        + 'a;
+
+impl<'a, A, R> kernel::hil::rubble::RubbleLinkLayer<'a, A, TockRubble<'a, A, R>>
+    for LinkLayerWrapper<'a, A, R>
+where
+    A: time::Alarm<'a>,
+    R: hil::rubble::radio::RubbleData<'a>
+        + ble_advertising::BleAdvertisementDriver<'a>
+        + ble_advertising::BleConfig
+        + 'a,
+{
+    type Error = rubble::Error;
+
+    fn new(device_address: hil::rubble::types::DeviceAddress, alarm: &'a A) -> Self {
+        LinkLayerWrapper(rubble::link::LinkLayer::new(
+            rubble::link::DeviceAddress::new(
+                device_address.bytes,
+                match device_address.kind {
+                    hil::rubble::types::AddressKind::Public => rubble::link::AddressKind::Public,
+                    hil::rubble::types::AddressKind::Random => rubble::link::AddressKind::Random,
+                },
+            ),
+            timer_wrapper::TimerWrapper::new(alarm),
+        ))
+    }
+
+    fn start_advertise(
+        &mut self,
+        interval: hil::rubble::types::Duration,
+        data: &[u8],
+        transmitter: &mut BleRadioWrapper<'a, R>,
+        tx: refcell_packet_queue::RefCellConsumer<'static>,
+        rx: refcell_packet_queue::RefCellProducer<'static>,
+    ) -> Result<hil::rubble::types::NextUpdate, Self::Error> {
+        // This is inefficient, as we're parsing then directly turning back into
+        // bytes. However, this is a minimal-effort code to comply with Rubble's
+        // interface.
+
+        // A more efficient method might be to write a more minimal parser, and
+        // just directly create `AdStructure::Unknown` instances.
+        let mut ad_byte_reader = rubble::bytes::ByteReader::new(data);
+
+        // Note: advertising data is at most 31 octets, and each ad must be at
+        // least 2 octets, so we have at most 16 adstructures.
+        // TODO: allocating this on the stack might not be the most wise.
+        let mut ad_structures =
+            [rubble::link::ad_structure::AdStructure::Unknown { ty: 0, data: &[] }; 16];
+        let mut num_ads = 0;
+        while ad_byte_reader.bytes_left() > 0 {
+            let ad = rubble::link::ad_structure::AdStructure::from_bytes(&mut ad_byte_reader)?;
+            ad_structures[num_ads] = ad;
+            num_ads += 1;
+            if num_ads > 16 {
+                return Err(rubble::Error::InvalidLength);
+            }
+        }
+        let ad_structures = &ad_structures[..num_ads];
+
+        let res = self.0.start_advertise(
```

Comment:
This doesn't work on android if the device is only advertising without supporting connections.

https://github.com/jonas-schievink/rubble/blob/6518ff8e44b45a70b85468c6ae330848a2347b71/rubble/src/link/mod.rs#L259-L280

The `start_advertise()` function directly calls `PduBuf::discoverable()` which is the wrong type since we don't respond to scan requests.

Rubble only seems to expose that one start advertising function, so I'm not sure how to fix this.

Reply:
Just a note: it's likely the scan responses that need to be supported. The android scanning application wouldn't be able to notice if a transmission was lying about supporting connections.

Reply:
Just to confirm, do you mean support? https://github.com/jonas-schievink/rubble/blob/ed59fd68d318fa7a6febbcbc6425463db37c6a05/rubble/src/link/advertising.rs#L617

Reply:
I think he means this: https://github.com/jonas-schievink/rubble/blob/ed59fd68d318fa7a6febbcbc6425463db37c6a05/rubble/src/link/advertising.rs#L624